### ISSUE TITLE
The global retryStrategy is only for Pod type (not Step/DAG template)

### ISSUE THREAD

**Issue by shuangkun** [argoproj/argo-workflows#13990]
# Summary

I hope to configure the global retryStrategy uniformly, but I don't want the Step/DAG type retry to cause the task to be re-run.

## Use Cases

Very complex workflow with unified retryStrategy, do not want to configure retryStrategy in each template.
Do not want to configure memoize, which is troublesome for developers.
Do not want the retry of Steps/DAG type template to cause the completed tasks to be re-run.

---

**Message from the maintainers**:

Love this feature request? Give it a 👍. We prioritise the proposals with the most 👍.

**Comment by jswxstw** [comment]
Completely agree! Setting `retryStrategy` uniformly in `templateDefaults` is simply a disaster: https://github.com/argoproj/argo-workflows/issues/13297#issuecomment-2203012122.

**Comment by tczhao** [comment]
Maybe we can filter node type in retry expression

**Comment by shuangkun** [comment]
> Maybe we can filter node type in retry expression

Yes, I tried to do that once, but couldn't find a variable to use directly. https://argo-workflows.readthedocs.io/en/latest/variables/ Maybe we need to add a variable to each template to identify which template it is.
Another way is to add a field in the retryStratepy at the workflow level, such as retryNodeType: [Steps, DAG, Container, Script, HTTP...]

**Comment by jswxstw** [comment]
> Maybe we need to add a variable to each template to identify which template it is.

https://github.com/argoproj/argo-workflows/blob/43c6abd1fe009f36c290545655080eb9c9c197fd/workflow/controller/operator.go#L2098
It seems that we only need to add a filtering condition here? Or should we add a new field or add a new retry policy in `retryStrategy` to control this?
```golang
processedTmpl.IsPodType() && woc.retryStrategy(processedTmpl) != nil
```

**Comment by shuangkun** [comment]
Yes. Whether adding filtering conditions or adding new fields, there needs to be a condition for the user to trigger. Either use expression or use retryNodeType configuration.

**Comment by jswxstw** [comment]
> Completely agree! Setting `retryStrategy` uniformly in `templateDefaults` is simply a disaster: [#13297 (comment)](https://github.com/argoproj/argo-workflows/issues/13297#issuecomment-2203012122).

The global `retryStrategy` also will cause the entire workflow topology and state to become inconsistent when hooks are used, see https://github.com/argoproj/argo-workflows/issues/13239#issuecomment-2463115971.
